src/src/PluginStructs/P014_data_struct.cpp

```diff
@@ -1,144 +1,218 @@
-  bool success = false;
 
-  switch (state) {
-    case SI7021_state::Uninitialized:
-    {
-      addLog(LOG_LEVEL_INFO, F("SI7021 : sensor not initialized !"));
-      init();
-      break;
-    }
-    case SI7021_state::Initialized:
-    {
-      // No call made to start a new reading
-      // Should be handled in the loop()
-      addLog(LOG_LEVEL_INFO, F("SI7021 : No read started !"));
-      break;
-    }
-    case SI7021_state::Wait_for_temperature_samples:
-    case SI7021_state::Wait_for_humidity_samples:
-    {
-      // Still waiting for data
-      // Should be handled in the loop()
-      addLog(LOG_LEVEL_ERROR, F("SI7021 : Read Error !"));
-      break;
-    }
 
-    case SI7021_state::New_values:
-    {
-      temperature = si7021_temperature / 100.0f;
-      humidity    = si7021_humidity / 10.0f;
-      state       = SI7021_state::Values_read;
-      success     = true;
-
-      if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-        String log = F("SI7021 : Temperature: ");
-        log += temperature;
-        addLogMove(LOG_LEVEL_INFO, log);
-        log  = F("SI7021 : Humidity: ");
-        log += humidity;
-        addLogMove(LOG_LEVEL_INFO, log);
-      }
-      break;
-    }
-    case SI7021_state::Values_read:
-    {
-      // This must be done in a separate call to
-      // make sure we only start reading when the plugin wants us to perform a reading.
-      // Change state of sensor for non blocking reading
-      state = SI7021_state::Initialized;
-      break;
-    }
-  }
-  return success;
+//The main state machine
+//Only perform the measurements with big interval to prevent the sensor from warming up.
+//returns true when the state changes and false otherwise
+bool P014_data_struct::update(uint8_t i2caddr, uint8_t resolution, uint8_t filter_power) {
+  //const unsigned long current_time = millis();
+  //int8_t ret=0;
+  
+  switch(state){
+    case P014_state::Uninitialized:
+        //make sure we do not try to initialize too often
+        //someone could enable the plugin but does not have the hardware connected
+        if (!timeOutReached(last_measurement_time + errCount*SI70xx_INIT_DELAY)) {
+          return false;
+        }
+
+        //we have to stop trying after a while
+        if (errCount>P014_MAX_RETRY){
+            state = P014_state::Error;
+            return true;
+        }
+
+        if (I2C_wakeup(i2caddr)!=0){
+          String log = F("SI70xx : Not available at address: ");
+          log += String(i2caddr, HEX);
+          addLog(LOG_LEVEL_ERROR, log);
+          errCount++;
+          last_measurement_time = millis();
+          return false;
+        }
+
+        if (init(i2caddr, resolution)){
+          state                 = P014_state::Initialized;
+          last_measurement_time = millis();
+          errCount              = 0;
+          return true;
+        }
+
+        last_measurement_time = millis();
+        errCount++;
+        return false;
+    break;
```

Comment:
`break` after a `return` is unreachable code.

Reply:
I put the break for completeness  the compiler should just exclude it as far as I know.

Reply:
I know that PVS Studio (which I use every now and then to check the code base) does warn with exactly that message on similar code.

Reply:
done. commented those out.